A59-audit-logging.md

```diff
@@ -0,0 +1,281 @@
+|NONE                  |NONE               |NONE                 |
+|ON_DENY               |ON_DENY            |ON_DENY              |
+|ON_ALLOW              |NONE               |ON_ALLOW             |
+|ON_DENY_AND_ALLOW     |ON_DENY            |ON_DENY_AND_ALLOW    |
+
+Again this is a subset of what users can technically do with xDS RBACs in the way
+described in the earlier section. But we think this represents the most common
+use cases.
+
+### Language Specific APIs
+
+_This section is to be done once the language-agnostic APIs are more finalized._
+
+_What is certain for now is we need to support third-party audit logger
+implementations registered by users._
+
+## Rationale
+
+So far, we have considered alternatives in a number of othorgonal aspects.
+
+### Utilizing the Access Log
+
+The [information](https://github.com/envoyproxy/envoy/blob/main/api/envoy/data/accesslog/v3/accesslog.proto)
+present in the existing access log is sufficient for audit purposes. However,
+access log does not happen till the response is sent and therefore does not
+fit the audit logging use case well. For example, for long-living RPCs, users
+would want to audit the event right after the authorization is enfoced but
+access log won't necessarily happen till hours later.
+
+### Audit Condition in the HTTP Connection Manager
+
+As an HTTP filter chain can contain multiple RBAC filters in general. The typical
+behavior that users will expect if they want to audit on allow is audit should
+only happen once after the evaluation of the last RBAC filter. From this
+perspective, the audit condition could be considered as something on top of all
+RBAC filters and thus configured in the [HTTP Connection Manager][HttpConnectionManager proto].
+
+However, given that xDS users normally do not craft RBACs on their own but instead
+rely on the control plane APIs, such as Istio Authorization Policy, it is
+reasonable to let the control plane handle the translation properly.
+Meanwhile, configuring audit condition in individual RBACs offers the flexibility
+when users really want to audit events related to a particular RBAC.
+
+### Audit Log Configured In gRPC Bootstrap
+
+Instead of having the audit log configuration delivered from the control plane,
+we could follow the xDS TLS certificate provider in [A29: xDS-Based Security for gRPC Clients and Servers](A29-xds-tls-security.md)
+where the configuration is placed in the gRPC bootstrap file.
+
+The advantage of that apporoach is the control plane would not need to know
+the platform-specific configuration options in the data plane, if any.
+But we concluded that in audit logging, the logger configuration should
+generally be uniform across the data plane and thus centralizing the configuration
+makes more sense. Moreover, the existing access log follows such a pattern
+in xDS, albeit not currently supported by gRPC.
+
+## Implementation
+
+The implementation order will be C++, Go, Java and then wrapped languages.
```

Comment:
Would it make sense to change it to 
"C++, Go, and then Java and wrapped languages"?

Reply:
You could emphasize C++ & Go with "Implementation will first be done for C++ and Go.  Following that will be Java, and finally wrapped languages will be implemented."

Reply:
Thank you, Larry! Done.